File: threads/src/main/java/im/threads/internal/utils/FileDownloader.java
package im.threads.internal.utils;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;
import java.util.Locale;

import javax.net.ssl.HttpsURLConnection;

import im.threads.internal.Config;

public class FileDownloader {
    private static final String TAG = "FileDownloader ";
    private final String path;
    private final File outputFile;
    private final DownloadLister downloadLister;

    private boolean isStopped;

    public FileDownloader(@NonNull String path, @NonNull String fileName, @NonNull Context ctx, @Nullable DownloadLister downloadLister) {
        this.path = path;
        this.outputFile = new File(getDownloadDir(ctx), generateFileName(path, fileName));
        this.downloadLister = downloadLister;
    }

    public static File getDownloadDir(Context ctx) {
        return ctx.getFilesDir();
    }

    public static String generateFileName(@NonNull String path, @NonNull String fileName) {
        StringBuilder sb = new StringBuilder();
        sb.append(getFileName(fileName))
                .append("(")
                .append(Uri.parse(path).getLastPathSegment())
                .append(")");
        final String ext = getFileExtension(fileName);
        if (ext != null) {
            sb.append(ext);
        }
        return sb.toString();
    }

    private static String getFileExtension(final String path) {
        if (path != null && path.lastIndexOf('.') != -1) {
            return path.substring(path.lastIndexOf('.'));
        }
        return null;
    }

    private static String getFileName(String fileName) {
        if (fileName.lastIndexOf('.') != -1) {
            return fileName.substring(0, fileName.lastIndexOf('.'));
        }
        return fileName;
    }

    public void stop() {
        isStopped = true;
    }

    public Long getFileLength(HttpURLConnection urlConnection) {
        try {
            List<String> values = urlConnection.getHeaderFields().get("Content-Length");
            if (values != null && !values.isEmpty()) {
                return Long.parseLong((String) values.get(0));
            }
        } catch (Exception e) {
            ThreadsLogger.e(TAG, "download", e);
        }
        return null;
    }

    public void download() {
        try {
            URL url = new URL(this.path);
            boolean isHTTPS = url.getProtocol().toLowerCase(Locale.ROOT).equals("https");
            if (isHTTPS && Config.instance.sslSocketFactoryConfig != null) {
                HttpsURLConnection.setDefaultSSLSocketFactory(Config.instance.sslSocketFactoryConfig.getSslSocketFactory());
            }
            HttpURLConnection urlConnection;
            if (isHTTPS) {
                urlConnection = (HttpsURLConnection) url.openConnection();
            } else {
                urlConnection = (HttpURLConnection) url.openConnection();
            }
            try {
                urlConnection.setRequestMethod("GET");
                urlConnection.setRequestProperty("X-Ext-Client-ID", PrefUtils.getClientID());
                urlConnection.setDoOutput(false);
                urlConnection.setUseCaches(false);
                urlConnection.setDoInput(true);
                urlConnection.setConnectTimeout(60000);
                urlConnection.setReadTimeout(60000);
                if (isHTTPS && Config.instance.sslSocketFactoryConfig != null) {
                    ((HttpsURLConnection)urlConnection).setHostnameVerifier((hostname, session) -> true);
                }
                Long length = getFileLength(urlConnection);
                FileOutputStream fileOutputStream = new FileOutputStream(outputFile);
                InputStream in = new BufferedInputStream(urlConnection.getInputStream());
                int len1;
                long bytesReaded = 0;
                long lastReadTime = System.currentTimeMillis();
                final byte[] buffer = new byte[1024 * 8];
                while ((len1 = in.read(buffer)) > 0 && !isStopped) {
                    fileOutputStream.write(buffer, 0, len1);
                    bytesReaded += len1;
                    if (length != null && (System.currentTimeMillis() > (lastReadTime + 500))) {
                        int progress = (int) Math.floor((((double) bytesReaded) / ((double) length)) * 100.0);
                        lastReadTime = System.currentTimeMillis();
                        if (downloadLister != null) {
                            downloadLister.onProgress(progress);
                        }
                    }
                }
                fileOutputStream.flush();
                fileOutputStream.close();
                if (!isStopped) {
                    if (downloadLister != null) {
                        downloadLister.onComplete(outputFile);
                    }
                }
            } catch (Exception e) {
                ThreadsLogger.e(TAG, "1 ", e);
                if (downloadLister != null) {
                    downloadLister.onFileDownloadError(e);
                }
            } finally {
                urlConnection.disconnect();
            }
        } catch (Exception e) {
            ThreadsLogger.e(TAG, "2 ", e);
            if (downloadLister != null) {
                downloadLister.onFileDownloadError(e);
            }
        }
    }

    public interface DownloadLister {
        void onProgress(double progress);

        void onComplete(File file);

        void onFileDownloadError(Exception e);

    }
}

File: threads/src/main/java/im/threads/internal/utils/FileDownloader.kt
package im.threads.internal.utils

import android.content.Context
import android.net.Uri
import android.text.TextUtils
import im.threads.internal.Config
import im.threads.internal.domain.logger.LoggerEdna
import java.io.BufferedInputStream
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import javax.net.ssl.HttpsURLConnection
import kotlin.math.floor

const val DOWNLOAD_PROGRESS_DELTA_TIME_MILLIS = 100
const val MAX_DOWNLOAD_PROGRESS = 100
const val DELTA_DOWNLOAD_PROGRESS = 2

class FileDownloader(
    private val path: String,
    fileName: String,
    ctx: Context,
    private val downloadLister: DownloadLister
) {
    private val outputFile: File = File(
        getDownloadDir(ctx),
        generateFileName(
            path,
            fileName
        )
    )

    private var isStopped = false

    fun download() {
        try {
            val url = URL(path)
            Config.instance.sslSocketFactoryConfig?.let {
                HttpsURLConnection.setDefaultSSLSocketFactory(it.sslSocketFactory)
            }
            val urlConnection = if (url.protocol.equals("https", ignoreCase = true)) {
                url.openConnection() as HttpsURLConnection
            } else {
                url.openConnection() as HttpURLConnection
            }
            try {
                urlConnection.requestMethod = "GET"
                urlConnection.setRequestProperty("X-Ext-Client-ID", PrefUtils.clientID)
                if (!TextUtils.isEmpty(PrefUtils.authToken)) {
                    urlConnection.setRequestProperty("Authorization", PrefUtils.authToken)
                }
                if (!TextUtils.isEmpty(PrefUtils.authSchema)) {
                    urlConnection.setRequestProperty("X-Auth-Schema", PrefUtils.authSchema)
                }
                urlConnection.doOutput = false
                urlConnection.useCaches = false
                urlConnection.doInput = true
                urlConnection.connectTimeout = 60000
                urlConnection.readTimeout = 60000

                if (urlConnection is HttpsURLConnection) {
                    urlConnection.setHostnameVerifier { _, _ -> true }
                }

                val length = getFileLength(urlConnection)
                val fileOutputStream = FileOutputStream(outputFile)
                val `in`: InputStream = BufferedInputStream(urlConnection.inputStream)
                var tempLength: Int
                var bytesRead: Long = 0
                var lastReadTime = System.currentTimeMillis()
                var lastReadProgress = 0
                val buffer = ByteArray(1024 * 8)

                while (`in`.read(buffer).also { tempLength = it } > 0 && !isStopped) {
                    fileOutputStream.write(buffer, 0, tempLength)
                    bytesRead += tempLength.toLong()
                    if (System.currentTimeMillis() > lastReadTime + DOWNLOAD_PROGRESS_DELTA_TIME_MILLIS) {
                        length?.let {
                            val progress =
                                floor(bytesRead.toDouble() / it * MAX_DOWNLOAD_PROGRESS).toInt()
                            lastReadTime = System.currentTimeMillis()
                            downloadLister.onProgress(progress.toDouble())
                        } ?: run {
                            lastReadProgress += DELTA_DOWNLOAD_PROGRESS
                            if (lastReadProgress >= MAX_DOWNLOAD_PROGRESS) {
                                lastReadProgress = 0
                            }
                            downloadLister.onProgress(lastReadProgress.toDouble())
                        }
                    }
                }

                fileOutputStream.flush()
                fileOutputStream.close()

                if (!isStopped) {
                    downloadLister.onComplete(outputFile)
                }
            } catch (e: Exception) {
                LoggerEdna.error("1 ", e)
                downloadLister.onFileDownloadError(e)
            } finally {
                urlConnection.disconnect()
            }
        } catch (e: Exception) {
            LoggerEdna.error("2 ", e)
            downloadLister.onFileDownloadError(e)
        }
    }

    fun stop() {
        isStopped = true
    }

    private fun getFileLength(urlConnection: HttpURLConnection): Long? {
        try {
            val values = urlConnection.headerFields["Content-Length"]
            if (values != null && values.isNotEmpty()) {
                return values[0]?.toLong()
            }
        } catch (e: Exception) {
            LoggerEdna.error("download", e)
        }
        return null
    }

    interface DownloadLister {
        fun onProgress(progress: Double)
        fun onComplete(file: File)
        fun onFileDownloadError(e: Exception?)
    }

    companion object {
        @JvmStatic
        fun getDownloadDir(ctx: Context): File {
            return ctx.filesDir
        }

        fun generateFileName(path: String, fileName: String): String {
            val sb = StringBuilder()
            sb.append(getFileName(fileName))
                .append("(")
                .append(Uri.parse(path).lastPathSegment)
                .append(")")
            val ext = getFileExtension(fileName)
            if (ext != null) {
                sb.append(ext)
            }
            return sb.toString()
        }

        private fun getFileExtension(path: String?): String? {
            return if (path != null && path.lastIndexOf('.') != -1) {
                path.substring(path.lastIndexOf('.'))
            } else null
        }

        private fun getFileName(fileName: String): String {
            return if (fileName.lastIndexOf('.') != -1) {
                fileName.substring(0, fileName.lastIndexOf('.'))
            } else fileName
        }
    }
}
